src/dune_rules/pkg_rules.ml

```diff
@@ -1149,11 +1458,7 @@ module DB = struct
 end
 
 module rec Resolve : sig
-  val resolve
-    :  DB.t
-    -> Context_name.t
-    -> Loc.t * Package.Name.t
-    -> [ `Inside_lock_dir of Pkg.t | `System_provided ] Memo.t
+  val resolve : DB.t -> Context_name.t -> Loc.t * Package.Name.t -> Pkg.t Memo.t
```

Comment:
I don't really see why this needed to change, but it doesn't seem like an improvement. An explanation on why this was needed would be helpful.

Reply:
Copied from the commit message of https://github.com/ocaml/dune/pull/10719/commits/48ed844f53629999193121d116c7db4de752775f:

The value of the PATH environment variable when building a pacage is
set such that all the executables in that package's dependency cone
are visible. This includes the ocaml compiler itself. In the past,
system-provided packages (such as dune) were omitted from this
dependency cone, along with their transitive dependencies.

This creates a problem for packages that depend on dune and don't
directly depend on the compiler. Such packages typically build with
`dune build -p ...`, and dune assumes that it's being run in an
environment where the ocamlc compiler is visible. Although the dune
package depends on the compiler, due to its exclusion from the
dependency cone used to set the PATH variable dune is unable to find
the compiler when building the package.

We haven't noticed this problem until now because it's rare to use
dune in environments where the compiler is not globally installed or
present in the active opam switch, however running dune on systems
without a global ocaml installation is something we're trying to
enable in the near future.

The fix is to add all packages, system-provided or not, to the graph
of [Pkg.t] objects that is used to compute the build environment for
each package. To prevent dune from attempting to build system-provided
packages, the check whether a package is system-provided is deferred
to the point where dune computes the list of packages to build.

Reply:
Thanks for the explanation. 

> This creates a problem for packages that depend on dune and don't directly depend on the compiler.

This points to an issue with such packages and not with dune. It is in fact possible to use dune without an OCaml compiler and we should encourage people to do so when appropriate (for building coq code for example). So I would much prefer if we could just fix the metadata for these packages in opam-repository rather than to add workarounds in dune.

If a workaround is truly necessary in dune, I would rather it be done in the least intrusive way possible. For example, by preprocessing all deps of such packages that depend on dune and not OCaml to also include OCaml.